change the type of generalBracket, fixes #63

gelisam opened the pull request:
I made the changes discussed in #63, added some tests, updated the documentation, and added a new `onError` method which, unlike `onException`, also runs its handler when the computation aborts which an error which is not an exception.

While we're making breaking changes, would it be worthwhile to also generalize the methods in `MonadCatch` so they can act on those non-exception errors?

RyanGlScott commented:
I've left some comments inline. I'd definitely like @snoyberg to look over this is as well.

> While we're making breaking changes, would it be worthwhile to also generalize the methods in `MonadCatch` so they can act on those non-exception errors?

Can you be more specific on what your intent is?

snoyberg commented on src/Control/Monad/Catch.hs:
```diff
@@ -653,44 +783,56 @@ catches a hs = a `catch` handler
 onException :: MonadCatch m => m a -> m b -> m a
 onException action handler = action `catchAll` \e -> handler >> throwM e
 
+-- | Run an action only if an error is thrown in the main action. Unlike
```
It may be worth spelling out that the difference is for transformers like `EitherT` and `MaybeT` that have multiple exit points.

snoyberg commented on src/Control/Monad/Catch.hs:
```diff
@@ -194,61 +214,95 @@ class MonadCatch m => MonadMask m where
     -> (a -> m b)
     -- ^ inner action to perform with the resource
-    -> m b
+    -> (a -> ExitCase b -> m c)
```
I should have commented on this earlier. I think it would make more sense to move the release action to before the use action, to match more closely the type signature of `bracket` in `base`.

snoyberg commented on src/Control/Monad/Catch.hs:
```diff
@@ -178,14 +199,13 @@ class MonadCatch m => MonadMask m where
-  -- functions are ignored.
+  -- | A generalized version of 'bracket' which uses 'ExitCase' to distinguish
+  -- the different exit cases, and returns the values of both the 'use' and
+  -- 'release' actions. You should probably use 'bracket' instead.
```
I'd probably clarify _why_ `bracket` is better to use (more ergonomic), and comment that many other functions (like `finally`) are built on top of this method.

snoyberg commented on src/Control/Monad/Catch.hs:
```diff
@@ -194,61 +214,95 @@ class MonadCatch m => MonadMask m where
-    -> m b
+    -> (a -> ExitCase b -> m c)
+    -- ^ release the resource, observing the outcome of the inner action
+    -> m (b, c)
```
I think this signature makes things much easier to understand.

gelisam commented:
>> While we're making breaking changes, would it be worthwhile to also generalize the methods in `MonadCatch` so they can act on those non-exception errors?
> 
> Can you be more specific on what your intent is?

With `MonadMask`, we can now handle two kinds of error cases: `ExitCaseException` and `ExitCaseAbort`. `MonadCatch` can only handle the `ExitCaseException` case. I thought it might be more consistent if `MonadCatch` was also extended to support both cases, by adding a method which could look like this, with `Nothing` representing the `ExitCaseAbort` case:

    generalCatch :: Exception e => m a -> (Maybe e -> m a) -> m a

But now that I'm thinking about this more, I don't think it makes a lot of sense to add a variant to the `catch` and `handle` family of functions which can catch the `ExitCaseAbort` case, because we don't have a concrete value to give to this handler. The user would be better-served by a monad-specific catching function like [`catchE`](https://hackage.haskell.org/package/transformers-0.5.5.0/docs/Control-Monad-Trans-Except.html#v:catchE).

gelisam commented:
> I'm not exactly sure what form your suggestion is going to take on

To clarify: I no longer have a suggestion for changing `MonadCatch`; I had one but I have retracted it.

gelisam commented:
I have now addressed all the inline comments. I have also added the an entry to the changelog about the new `onError` function, added comments around the `StateT` and `ExceptT` implementations as a reminder to keep the documentation's copy of their implementation in sync, and I have added a few more `@since` annotations.

RyanGlScott commented on src/Control/Monad/Catch.hs:
```diff
@@ -653,44 +796,61 @@ catches a hs = a `catch` handler
+-- example, if @f@ is an 'ExceptT' computation which aborts with a 'Left', the
+-- computation @onError f g@ will execute @g@, while @onException f g@ will not.
+--
+-- For monads which, unlike 'ExceptT' and 'MaybeT', only have a single exit
```
The phrasing of this sentence is a little awkward. I would prefer stating it like:

"This distinction is only meaningful for monads which have multiple exit points, such as 'Except' and 'MaybeT'. For monads that only have a single exit point, there is no difference between 'onException' and 'onError', except that 'onError' has a more constrained type."

RyanGlScott commented on src/Control/Monad/Catch.hs:
```diff
@@ -194,61 +217,97 @@ class MonadCatch m => MonadMask m where
     -> (a -> m b)
     -- ^ inner action to perform with the resource
-    -> m b
+    -> m (b, c)
```
How do we feel about the order of type variables in `generalBracket`? Our primary motivator for picking this order of arguments was to match what `bracket` does, but now we have:

```haskell
bracket
  :: IO a
  -> (a -> IO b)
  -> (a -> IO c)
  -> IO c

generalBracket
  :: m a
  -> (a -> ExitCase b -> m c)
  -> (a -> m b)
  -> m (b, c)
```

Now, the type variables `b` and `c` are in the opposite order as in `bracket`. We could conceivably change this to:

```haskell
generalBracket
  :: m a
  -> (a -> ExitCase c -> m b)
  -> (a -> m c)
  -> m (b, c)
```

Or even:

```haskell
generalBracket
  :: m a
  -> (a -> ExitCase c -> m b)
  -> (a -> m c)
  -> m (c, b)
```

I don't have a strong opinion one way or the other. Thoughts?

gelisam replied:
I don't have a strong opinion either. I'll change it to your first suggestion in order to at least match bracket's `(a -> m c)`, then change it back if someone feels strongly about this :)

RyanGlScott replied:
OK. The return type is interesting because it might hint at an order of evaluation. That is, the action which returns `c` runs before the action which returns `b`, hence `m (c, b)`. On the other hand, this puts us back into a state where the type variables are out-of-order again. Perhaps the order of evaluation is already obvious because `ExitCase c` appears as an argument to the function which returns an `m b`?

gelisam replied:
Now that I am attempting the change, I feel like changing the variable names is a bad idea. It's important to thread the state in the correct order: from acquire to use, then from use to release. By swapping the order of `use` and `release` to match `bracket`, we have already made it less obvious that this is what implementors need to do. Let's at least keep these variable names as is in order to give a hint of the right ordering.

gelisam replied:
Would it be crazy to change change the type variables in `bracket` to match `generalBracket`, to make it clearer that even though `use` is last so `bracket` curries better, it is executed in the middle of the `a` and `c` bits?

    bracket
     :: IO a
      -> (a -> IO c)
      -> (a -> IO b)
      -> IO b

RyanGlScott replied:
That sounds like a reasonable suggestion too (short of actually swapping the order of those arguments).

gelisam replied:
done

RyanGlScott commented:
Looking good! At this point, the only issue is that this doesn't build on old GHCs due to `<$>` and `<*>` not being in scope. I think you can fix both of these issues with:

```haskell
import Control.Applicative
```

At around line 106 or so (within the `#elif __GLASGOW_HASKELL__ < 710 ... #endif` CPP).

RyanGlScott commented:
So close :)

Travis is now complaining:

```
src/Control/Monad/Catch.hs:835:27:
    Could not deduce (Functor m) arising from a use of ‘fmap’
    from the context (MonadMask m)
      bound by the type signature for
                 bracket :: MonadMask m => m a -> (a -> m c) -> (a -> m b) -> m b
      at src/Control/Monad/Catch.hs:834:12-64
    Possible fix:
      add (Functor m) to the context of
        the type signature for
          bracket :: MonadMask m => m a -> (a -> m c) -> (a -> m b) -> m b
    In the first argument of ‘(.)’, namely ‘fmap fst’
    In the expression:
      fmap fst . generalBracket acquire (\ a _exitCase -> release a)
    In an equation for ‘bracket’:
        bracket acquire release
          = fmap fst . generalBracket acquire (\ a _exitCase -> release a)

src/Control/Monad/Catch.hs:852:34:
    Could not deduce (Functor m) arising from a use of ‘fmap’
    from the context (MonadMask m)
      bound by the type signature for
                 bracketOnError :: MonadMask m =>
                                   m a -> (a -> m c) -> (a -> m b) -> m b
      at src/Control/Monad/Catch.hs:851:19-71
    Possible fix:
      add (Functor m) to the context of
        the type signature for
          bracketOnError :: MonadMask m =>
                            m a -> (a -> m c) -> (a -> m b) -> m b
    In the first argument of ‘(.)’, namely ‘fmap fst’
    In the expression:
      fmap fst
      . generalBracket
          acquire
          (\ a exitCase
             -> case exitCase of {
                  ExitCaseSuccess _ -> return ()
                  _ -> do { _ <- release a;
                            .... } })
    In an equation for ‘bracketOnError’:
        bracketOnError acquire release
          = fmap fst
            . generalBracket
                acquire
                (\ a exitCase
                   -> case exitCase of {
                        ExitCaseSuccess _ -> return ()
                        _ -> do { ... } })
```

I think replacing these uses of `fmap` with `liftM` will do the trick.

RyanGlScott commented:
Oh, right. Currently, the `Control.Applicative` import is only in scope _between_ GHC 7.6 and 7.10, which is why it's now only failing on GHC 7.4. Try taking out that import and putting it in its own:

```haskell
#if __GLASGOW_HASKELL__ < 710
import Control.Applicative
#endif
```

gelisam commented:
Slowly but surely!

RyanGlScott commented:
Thanks, @gelisam! I'll try to get a new release out next week.
